kube-runtime/src/watcher.rs

```diff
@@ -154,6 +154,24 @@ struct FullObject<'a, K> {
     api: &'a Api<K>,
 }
 
+/// Configurable list semantics for `watcher` relists
+#[derive(Clone, Default, Debug, PartialEq)]
+pub enum ListSemantic {
+    /// List calls perform a full quorum read for most recent results
+    ///
+    /// Prefer this if you have strong consistency requirements. Note that this
+    /// is more taxing for the apiserver and can be less scalable for the cluster.
+    MostRecent,
+
+    /// List calls returns cached results from apiserver
+    ///
+    /// This is faster and much less taxing on the apiserver, but can result
+    /// in much older results than has previously observed for `Restarted` events,
+    /// particularly in HA configurations, due to partitions or stale caches.
+    #[default]
+    Cached,
```

Comment:
Don't like to be a party crasher, but this can be tricky to debug in some cases, and changes the default behavior. I just want to highlight that we can go with the most recent semantic in the upcoming release and reconsider the default in the future. Yet, of course, the final decision is up to you.

Reply:
Appreciate the input. I am not fully convinced changing the default here is the best way forward either. I'll wait for Nat to chime in.

Reply:
Having thought more about it, I have changed it in watcher as well so that we are not changing behaviour anywhere by default.

However, users can opt-in to the more performant, but less consistent via `Config::any_semantic`.

Reply:
Was going back and forth on this a bit, but am now at ease with not changing the default and having `MostRecent` - the more consistent one but heavier call - as a default.

While it maybe makes sense for Controller to have a less consistent default - as things there are generally eventually consistent and meant to deal with a few redundant reconciles anyway, it's a little more iffy for direct `watcher` use. Have emphasised this in the docs, for the `ListSemantic` enum. One sticking point for me is that, people like opting into performance a lot more than they like having to debug less consistency (if that ever happens), and this would be a hard to detect break.